contrib/win32/win32compat/w32-sshfileperm.c

```diff
@@ -0,0 +1,236 @@
+/*
+* Author: Yanbing Wang <yawang@microsoft.com>
+*
+* Copyright (c) 2009, 2011 NoMachine
+* All rights reserved
+*
+* Support file permission check functions' replacements needed to let the
+* software run on Win32 based operating systems.
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions
+* are met:
+*
+* 1. Redistributions of source code must retain the above copyright
+* notice, this list of conditions and the following disclaimer.
+* 2. Redistributions in binary form must reproduce the above copyright
+* notice, this list of conditions and the following disclaimer in the
+* documentation and/or other materials provided with the distribution.
+*
+* THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+* IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+* OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+* IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+* NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+* THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <Windows.h>
+#include <Sddl.h>
+#include <Aclapi.h>
+#include <Ntsecapi.h>
+#include <lm.h>
+
+#include "inc\pwd.h"
+#include "sshfileperm.h"
+#include "misc_internal.h"
+#include "debug.h"
+
+#define SSHD_ACCOUNT L"NT Service\\sshd"
+
+/*
+* To Check if the user have right owner and permission set.
+* Returns 0 on success and -1 on failure
+*/
+int
+secure_file_permission(const char *name, struct passwd * pw)
+{	
+	PSECURITY_DESCRIPTOR pSD = NULL;
+	PSID owner_sid = NULL, user_sid = NULL;
+	PACL dacl = NULL;
+	int ret = 0;
+
+	if (pw == NULL) {
+		debug3("invalid parameter pw is null");
+		errno = EINVAL;
+		return -1;
+	}	
+	if (ConvertStringSidToSid(pw->pw_sid, &user_sid) == FALSE ||
+		(IsValidSid(user_sid) == FALSE)) {
+		debug3("failed to retrieve the sid of the pwd");
+		ret = -1;
+		goto cleanup;
+	}
+
+	/*Get the owner sid of the file.*/
+	if ((ret = GetNamedSecurityInfo(name, SE_FILE_OBJECT,
+		OWNER_SECURITY_INFORMATION | DACL_SECURITY_INFORMATION,
+		&owner_sid, NULL, &dacl, NULL, &pSD)) != ERROR_SUCCESS ||
+		(IsValidSid(owner_sid) == FALSE) ||
+		(IsValidAcl(dacl) == FALSE)) {
+		debug3("failed to retrieve the owner sid and dacl of file %s with error code: %d", name, ret);
```

Comment:
the ret is the error code (https://msdn.microsoft.com/en-us/library/windows/desktop/aa446645(v=vs.85).aspx). I will use different var to be clear.
added var to log the return of IsValidSid / IsValidAc